custom-metrics-stackdriver-adapter/pkg/adapter/coreprovider/client_test.go

```diff
@@ -0,0 +1,326 @@
+	{
+		numOfPods:       534,
+		numOfContainers: 1,
+	},
+	{
+		numOfPods:       99,
+		numOfContainers: 5,
+	},
+	{
+		numOfPods:       100,
+		numOfContainers: 7,
+	},
+	{
+		numOfPods:       201,
+		numOfContainers: 3,
+	},
+}
+
+var nodetests = []struct {
+	numOfNodes int
+}{
+	{
+		numOfNodes: 1,
+	},
+	{
+		numOfNodes: 2,
+	},
+	{
+		numOfNodes: 15,
+	},
+	{
+		numOfNodes: 299,
+	},
+	{
+		numOfNodes: 300,
+	},
+	{
+		numOfNodes: 301,
+	},
+	{
+		numOfNodes: 132,
+	},
+	{
+		numOfNodes: 834,
+	},
+}
+
+func TestCoreprovider_getContainerMetric(t *testing.T) {
+	tr, _ :=
+		translator.NewFakeTranslator(2*time.Minute, time.Minute, "my-project", "my-cluster", "my-zone", time.Date(2017, 1, 2, 13, 2, 0, 0, time.UTC), true)
+
+	for _, testType := range testsTypes {
+		for _, tc := range podtests {
+			t.Run(fmt.Sprintf("%s%+v", testType, tc), func(t *testing.T) {
+				possibleResponses := make([]responseEntry, 0)
+				allPods := make([]string, 0)
+
+				for podId := 1; podId <= tc.numOfPods; podId++ {
+					for containerId := 1; containerId <= tc.numOfContainers; containerId++ {
+						var entry responseEntry
+						if testType.metric == "CPU" {
+							entry = getContainerGuageResponseEntry(podId, containerId)
+						} else {
+							entry = getContainerCumulativeResponseEntry(podId, containerId)
+						}
+						possibleResponses = append(possibleResponses, entry)
+						if containerId == 1 {
+							allPods = append(allPods, entry.resourceName)
+						}
+					}
+				}
+
+				client := newFakeClient(tr, possibleResponses)
+				var metricResult map[string]map[string]resource.Quantity
+				var timeResult map[string]api.TimeInfo
+				var err error
+
+				if testType.metric == "CPU" {
+					metricResult, timeResult, err = client.getContainerCPU(allPods)
+				} else {
+					metricResult, timeResult, err = client.getContainerRAM(allPods)
+				}
+				if err != nil {
+					t.Fatalf("%s", err)
+				}
+
+				for _, entry := range possibleResponses {
+					expectedVal, err := translator.GetQuantityValue(*entry.response.Points[0])
+					if err != nil {
+						t.Fatalf("Internal test error - can't get expected metric value")
+					}
+					namespaceName := entry.response.Resource.Labels["namespace_name"]
+					containerName := entry.response.Resource.Labels["container_name"]
+					key := namespaceName + ":" + entry.resourceName
+
+					if !reflect.DeepEqual(metricResult[key][containerName], *expectedVal) {
+						t.Errorf("Response for %v metric - expected: %v, but have %v", *entry.response.Resource, *expectedVal, metricResult[key][containerName])
+					}
+
+					expectedTime, err := time.Parse(time.RFC3339, entry.response.Points[0].Interval.EndTime)
+					if err != nil {
+						t.Fatalf("Internal Error, can't parse expected time")
+					}
+					if !reflect.DeepEqual(timeResult[key].Timestamp, expectedTime) {
+						t.Errorf("TimeInfo for %v metric - expected: %v, but have %v", *entry.response.Resource, expectedTime, timeResult[key].Timestamp)
+					}
+				}
```

Comment:
I don't understand what your testing. 

For code that splits list (all pods) into multiple lists (pods in single query) of size 100 I would expect something like:
* Test if length of requests is correct (all requests beside last one is size 100, and sum matches)
* Test if all pods are tested (throw pod names into map and check if size stays the same)

Reply:
Now I'm testing:
- If every pod was in request exactly one time
- If there is no request with more than 100 pods
- If the number of requests is minimal possible (equal to ceil(#pods / 100))
- If every container in included into result - it may be possible, that every pod was requested as required but it's result wasn't included in response - testing only requests don't cover all code

I think that better than require exact structure of request (100, 100, 100, ..., #pods % 100) It's better to check listed requirements. There could be another implementation which divides pod into equal-size groups and it's also a good implementation.